### ISSUE TITLE
Question: Performance difference between a Rayon for loop Vs Polars?

### ISSUE THREAD

**Issue by selvavm** [pola-rs/polars#2364]
Hi

I tried to port one of my old casebase to polars (was using Rayon and Ndarray earlier) to see if there is a performance improvements. For some strange reason, I am seeing reduced performance. 

Since my codebase is quite huge with lot of other crates, I will try to explain,

My code with Ndarray + Rayon,

```
rayon::ThreadPoolBuilder::new().num_threads(8).build_global().unwrap();
file_names.par_iter().for_each( |file_name| {
    //read file as ndarray
   // Calculate aggregation (sum, max, mean....)
   sled.insert(file, MyStatistics);
}
```

With polars,
```
rayon::ThreadPoolBuilder::new().num_threads(8).build_global().unwrap();
let lf = LazyFrame::scan_parquet(
        format!("{}\\p\\*.parquet", file_path),
        Default::default()
    ).unwrap();
let column = col("a");
let df = lf=
        .groupby([col("file_name")])
        .agg([
            column.clone().max().alias("max"),
            column.clone().min().alias("min"),
            column.clone().mean().alias("mean"),
            column.clone().std().alias("std"),
        ])
        .collect();
sled.insert(file, MyStatistics);
```

Below are the statistics,

| File Count | NDarray + Rayon| Polars |
| ------------- |:-------------:| -----:|
| 1500 files      | 0.7s |   150s|
| 1 file              | 0.4ms | 12ms|

I tried changing allocator as specified in the doc

Questions:

- Whatever thread count I change, there is no difference in polars performance. But, in my Ndarray code, I see huge drop in performance (due to blocking or unused threads)
- I always know the file schema is same and same column order. Does polars trying to derive this?
- When I have single file, the performance improved (but not as same as Ndarray code). Could it be due to glob support being relatively new?

**Comment by ritchie46** [comment]
Not really a fair comparisson. The polars glob reads all files reallocates them to a single dataframe and then you do a groupby over all those rows.

Whereas the ndrray query is written embarrassingly parallel.

**Comment by ritchie46** [comment]
Try writing the polars query similar to what you do wth ndarray.

**Comment by jorgecarleitao** [comment]
The queries seem to answer different questions - latter is a group by over the whole dataset, the first is an aggregation across partitions. They won't yield the same result.

**Comment by ritchie46** [comment]
Try something like this:

```rust
   let dfs_iter = file_names.par_iter().map( |file_name| {
        //read file as ndarray
        // Calculate aggregation (sum, max, mean....)

        let column = col("a");
        LazyFrame::scan_parquet(
            format!("{}\\p\\*.parquet", file_path),
            Default::default()
        ).select([
            column.clone().max().alias("max"),
            column.clone().min().alias("min"),
            column.clone().mean().alias("mean"),
            column.clone().std().alias("std"),
        ]).collect().unwrap()
    });
    concat_df(dfs_iter)
```

**Comment by selvavm** [comment]
Thanks all for comments. I am new to `Dataframe` concepts and still learning it. 

@ritchie46 - The script you shared was the first approach I tried while porting, I was (and I am) getting `thread '<unknown>' has overflowed its stack` and thus, I thought we cannot use `Dataframe` inside `par_iter`. I was under an assumption, `Dataframe`  takes care of threading and so I should not do my own.

**Comment by ritchie46** [comment]
And if you install it on the same threadpool?

This is what we do internally in polars for `collect_all`. Which is able to parallelize multiple `LazyFrames`:

```rust
        polars_core::POOL.install(|| {
            lfs.par_iter()
                .map(|lf| {
                    let df = lf.ldf.clone().collect()?;
                })
                .collect::<polars_core::error::Result<Vec<_>>>()
        })
```

**Comment by ritchie46** [comment]
> I was under an assumption, Dataframe takes care of threading and so I should not do my own.

You generally don't have to, but now you want to parallelize over multiple inputs instead of over operations with a query. Which is different.

I've added this utility to make it easier for you:

https://github.com/pola-rs/polars/pull/2371

You can just add a `Vec<LazyFrame>` and this will produce a  `Vec<DataFrame>`.

**Comment by selvavm** [comment]
Thanks @ritchie46. It is quite useful. 

I think my bottleneck is happening while doing the `filter_by(file_name_col)`. If I loop files outside, the performance is quite similar because I don't have `group_by` and `filter` in my plan.

I hope one day we will have something similar to `filter_by_partition(filename)` as proposed in https://github.com/pola-rs/polars/issues/2257 (wishful thinking). Then, `Polars` wont have to look through each rows but just each partition (in my case file_name). I am not sure if many will be interested in this kind of feature but I do :). 

I tried to see if I can modify the code but then figured out I am not really that good in `Rust` yet. I could see that `scan_parquet` is looping through each files while initialization itself, so, if I change, it will affect lot of places.

Side Note: I tried to export the `POOL` by modifying `polars::prelude` and tried. It works and I didnt get the `stackoverflow`. 



as you suggested the performance is on par with `NDaaray
